### PR TITLE
Headless Chrome update

### PR DESCRIPTION
**Pull request by ebidel**
R: @petele @drufball @PaulKinlan 

- [x] Tech review 
- [x] Tech writer review

### CODE REVIEW

**Comment by ebidel** [comment]
@jpmedley @kaycebasques can one of you take a look? Content is stable enough to review.

**Comment by kaycebasques** [comment]
On it!

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 1]
```diff
@@ -0,0 +1,369 @@
+project_path: /web/_project.yaml
+book_path: /web/updates/_book.yaml
+description: Getting started with Headless Chrome
+
+{# wf_updated_on: 2017-04-27 #}
+{# wf_published_on: 2017-04-27 #}
+{# wf_tags: chrome59,headless,testing #}
+{# wf_featured_image: /web/updates/images/generic/chrome-devtools.png #}
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: You want to use the Chrome DevTools logo?

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+{# wf_featured_snippet: Learn about Headless Chrome, shipping in Chrome 59. #}
+
+# Getting started with Headless Chrome {: .page-title }
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: s/started/Started/

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 3]
```diff
@@ -0,0 +1,369 @@
+project_path: /web/_project.yaml
+book_path: /web/updates/_book.yaml
+description: Getting started with Headless Chrome
+
+{# wf_updated_on: 2017-04-27 #}
+{# wf_published_on: 2017-04-27 #}
+{# wf_tags: chrome59,headless,testing #}
+{# wf_featured_image: /web/updates/images/generic/chrome-devtools.png #}
+{# wf_featured_snippet: Learn about Headless Chrome, shipping in Chrome 59. #}
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: For the featured snippet maybe you want to hint at why / how peeps use Headless Chrome?

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+<style>
+figure {
+  text-align: center;
+}
+</style>
+
+### TL;DR {: #tldr .hide-from-toc}
+
+[Headless Chrome](https://chromium.googlesource.com/chromium/src/+/lkgr/headless/README.md)
+is a way to run the Chromium browser in a headless environment. Essentially, running Chrome without chrome! It brings **all modern web platform features** provided
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: clarify difference b/t Chromium and Chrome? Maybe not here, but somewhere in the doc...

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+by Chromium and the Blink rendering engine to the command line.
+
+Why is that useful?
+
+A headless browser is great tool for automated testing and server environments where you
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: s/is great/is a great/

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+don't need a visible UI shell. For example, you may want to run some tests against
+a real web page, create a PDF of it, or just inspect how the browser renders an URL.
+
+Headless mode is available on Mac and Linux in Chrome 59. Windows support
+is coming soon!
+{: .note }
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: Remove this and add `Note:` at the beginning. It'll get formatted automagically

Note: Headless mode is available...

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+
+## Starting Headless (CLI) {: #cli }
+
+The easiest way to get started with headless mode is to open the Chrome binary
+from the command line. If you've got Chrome 59+ installed, start Chrome with the `--headless` flag:
+
+    chrome \
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: Is this an alias?

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+      --headless \                   # Runs Chrome in headless mode.
+      --disable-gpu \                # Temporarily needed for now.
+      --remote-debugging-port=9222 \
+      https://www.chromestatus.com   # URL to open. Defaults to about:blank.
+
+**Note**: Right now, you'll also want to include the `--disable-gpu` flag.
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: Plz use standard note format... see above

Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: Add a sentence explaining why `disable-gpu` is needed?

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+That will eventually go away.
+{: .note }
+
+`chrome` should point to your installation of Chrome. The exact location will
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: oh, ok

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+    alias chromium="/Applications/Chromium.app/Contents/MacOS/Chromium"
+
+## Command line features {: features }
+
+In some cases, you may not need to [programmatically script](#node) Headless Chrome.
+There are some [useful command line flags](https://cs.chromium.org/chromium/src/headless/app/headless_shell_switches.cc)
+to perform common tasks.
+
+**Note**: You may also need to include the `--disable-gpu` flag for now when
+running these commands
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: End with a period

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+    alias chrome-canary="/Applications/Google\ Chrome\ Canary.app/Contents/MacOS/Google\ Chrome\ Canary"
+    alias chromium="/Applications/Chromium.app/Contents/MacOS/Chromium"
+
+## Command line features {: features }
+
+In some cases, you may not need to [programmatically script](#node) Headless Chrome.
+There are some [useful command line flags](https://cs.chromium.org/chromium/src/headless/app/headless_shell_switches.cc)
+to perform common tasks.
+
+**Note**: You may also need to include the `--disable-gpu` flag for now when
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: porque, amigo

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+
+### Create a PDF {: dom }
+
+The `--print-to-pdf` flag creates a PDF of the page:
+
+    chrome --headless --print-to-pdf https://www.chromestatus.com/
+
+### Taking screenshots {: #screenshots }
+
+To capture the screenshot of a page, use the `--screenshot` flag:
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: s/capture the/capture a/

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+
+In some cases, you may not need to [programmatically script](#node) Headless Chrome.
+There are some [useful command line flags](https://cs.chromium.org/chromium/src/headless/app/headless_shell_switches.cc)
+to perform common tasks.
+
+**Note**: You may also need to include the `--disable-gpu` flag for now when
+running these commands
+{: .note }
+
+### Printing the DOM {: dom }
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: You could probably move these sections to a single "Useful flags" section and then just list them all in a table... would be faster to digest

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+    # Size of a standard letterhead.
+    chrome --headless --screenshot --window-size=1280,1696 https://www.chromestatus.com/
+
+    # Nexus 5x
+    chrome --headless --screenshot --window-size=412,732 https://www.chromestatus.com/
+
+Running with `--screenshot` will produce a file named `screenshot.png` in the
+current working directory. If you're looking for full page screenshots, things
+are a tad more involved. There's a great blog
+post from David Schnurr that has you covered. Check out "[Using headless Chrome as an automated screenshot tool
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: Remove the quotes from the link

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+](https://medium.com/@dschnr/using-headless-chrome-as-an-automated-screenshot-tool-4b07dffba79a)".
+
+
+## Debugging Chrome without a browser UI? {: #frontend }
+
+When you run Chrome with `--remote-debugging-port=9222`, it starts an instance
+with the [DevTools Protocol][dtprotocol] enabled. The
+protocol is what we use to communicate with Chrome and drive the headless
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: Who's "we"? Rephrase to something more explicit, plz

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+in another browser to check that everything is working. You'll see a list of
+inspectable pages where you can click through and see what Headless is rendering:
+
+<figure>
+  <img src="/web/updates/images/2017/04/headless-chrome/remote-debugging-ui.png"
+       class="screenshot" alt="DevTools Remote ">
+  <figcaption>DevTools remote debugging UI</figcaption>
+</figure>
+
+From here, you can use the familiar DevTools features to inspect, debug, and tweak
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: oxford comma makes me happy, thank u

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 17]
```diff
@@ -0,0 +1,369 @@
...
+In the previous section, we [started Chrome manually](#cli) using `--headless --remote-debugging-port=9222`. However, to fully automate tests, you'll probably
+want to spawn Chrome _from_ your application.
+
+One way is to use `child_process`:
+
+```javascript
+const exec = require('child_process').exec;
+
+function launchHeadlessChrome(url, callback) {
+  const CHROME = '/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome';
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: add comment that you're assuming mac?

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 18]
```diff
@@ -0,0 +1,369 @@
...
+  exec(`${CHROME} --headless --remote-debugging-port=9222 ${url}`, callback);
+}
+
+launchHeadlessChrome('https://www.chromestatus.com', (err, stdout, stderr) => {
+  ...
+});
+```
+
+But things get tricky if you want a portable solution that works across multiple
+platforms. Just look at that hard coded the path to Chrome :(
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: s/hard coded the path/hard-coded path/

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 19]
```diff
@@ -0,0 +1,369 @@
...
+
+#### Using Lighthouse's ChromeLauncher {: #nodechromelauncher }
+
+[Lighthouse](/web/tools/lighthouse/) is a marvelous
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: so objective ;)

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 20]
```diff
@@ -0,0 +1,369 @@
...
+tool for testing the quality of your web apps. One thing people don't realize
+is that it ships with some really nice helper modules for working with Chrome.
+One of those modules is `ChromeLauncher`. `ChromeLauncher` will find where
+Chrome is installed, set up a debug instance, launch the browser, and kill it
+when your program is done. Best part is that it works cross-platform thanks to
+Node!
+
+The Lighthouse team is exploring a standalone package for `ChromeLauncher` with
+an improved API. Let us know if you have [feedback](https://github.com/GoogleChrome/lighthouse/issues/2092).
+{: .note }
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: use std note format

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 21]
```diff
@@ -0,0 +1,369 @@
...
+Chrome fire up in the task manager that loaded `about:blank`. Remember, there
+won't be any browser UI. We're headless.
+
+To control the browser, we need the DevTools protocol!
+
+### Retrieving information about the page {: #useprotocol }
+
+[chrome-remote-interface](https://www.npmjs.com/package/chrome-remote-interface)
+is a great Node package that provides high-level APIs on top of the
+[DevTools Protocol][dtprotocol]. You can use it orchestrate Headless
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: s/use it orchestrate/use it to orchestrate/

File: src/content/en/updates/2017/04/headless-chrome.md [hunk 22]
```diff
@@ -0,0 +1,369 @@
...
+Chrome, navigate to pages, and fetch information about those pages.
+
+The DevTools protocol can do a ton of interesting stuff, but it can be a bit
+daunting at first. I recommend spending a bit of time browsing the [DevTools Protocol API Viewer][dtviewer], first. Then, move on to the `chrome-remote-interface` API docs to
+see how it wraps the raw protocol.
+{: .warning }
```
Comment (kaycebasques) [inline src/content/en/updates/2017/04/headless-chrome.md]: use the std format, same as before with `Note:`

File: src/content/en/updates/2017/04/headless-chrome.md (hunk 7 above)
Comment (ebidel) [inline src/content/en/updates/2017/04/headless-chrome.md]: yea, from previous section

File: src/content/en/updates/2017/04/headless-chrome.md (hunk 8 above)
Comment (ebidel) [inline src/content/en/updates/2017/04/headless-chrome.md]: Bugs basically. Don't it's worth calling out too much more than this.

File: src/content/en/updates/2017/04/headless-chrome.md (hunk 13 above)
Comment (ebidel) [inline src/content/en/updates/2017/04/headless-chrome.md]: It's under a "useful command line feature" heading. I only wanted to call out a few.

**Comment by ebidel** [comment]
Thanks @kaycebasques. PTAL. Think I got everything.

**Comment by ebidel** [comment]
@kaycebasques can i get your blessing. This needs to go out for early Monday.

**Comment by kosamari** [comment]
going to merge this. if needs edit, I'm here to deploy later.

**Pull request merged**